### PR TITLE
GET, POST, DELETE method for /daemons route

### PR DESCRIPTION
**Pull request by swhsiang**
# WHAT / WHY

* Implemt GET, POST, DELETE method for /daemons route
* Implement GET method for /cluster route
* Move the instance of ecs_state to a global variable. (same as session.Session, ecs.New())
* Move the cluster variable from local variable to a global variable
* Change the input parameter of StartTask function from (cluster string, taskDefinitions []string) to (taskDefinitions []string)

However, It is not fully functional. (Lacks a job scheduler to check the status of the host or register task on each host every 5 mins) I want to divide the implementation of this feature into two parts. Thus, I create this pr. 

Spec
```
GET /daemons

response:
{
error: true/false,
data: ""
}

POST /daemons
{
error: true/false
}

status code: 202 Accepted, for asyc request.

DELETE /daemons
{
  error: true/false,
  task-definition-arn: ""
}
```

About issue #11

### CODE REVIEW

**Comment by tnachen** [comment]
Fix your commit message (Implemt -> Implement)

File: scheduler.go [hunk 1]
```diff
@@ -17,6 +15,31 @@ import (
...
+
+	// State instance of ecs_state
+	State ecs_state.StateOps
+
+	// Cluster the name of cluster
+	Cluster string
+)
+
+func initialize(cluster string, awsRegion string) {
+
```
Comment (tnachen) [inline scheduler.go]: Remove the extra line

File: scheduler.go [hunk 2]
```diff
@@ -17,6 +15,31 @@ import (
...
+	Sess, _ = session.NewSession(&aws.Config{Region: aws.String(awsRegion)})
+	Client = ecs.New(Sess)
+	State = ecs_state.Initialize(cluster, Client)
+	State.RefreshClusterState()
+	State.RefreshContainerInstanceState()
+	State.RefreshTaskState()
+
```
Comment (tnachen) [inline scheduler.go]: Remove the extra line

File: api.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+func postDaemonHandler(c *gin.Context) {
+	taskDefinition := c.PostForm("taskDefinition")
+
+	c.JSON(http.StatusAccepted, gin.H{
+		"error": false,
+	})
+
+	// TODO Implement the function that register a batch job to lunch task on each host every 5 mins
```
Comment (tnachen) [inline api.go]: lunch -> launch.
And we shouldn't just go launch tasks on each host every 5 minutes.
We should listen to changes to the cluster and do actions accordingly in a reactive manner.

File: scheduler.go [hunk 4]
```diff
@@ -38,19 +61,10 @@ func selectUnlaunchedInstances(instances []ecs_state.ContainerInstance, newTask
...
-	client := ecs.New(sess)
-
-	state := ecs_state.Initialize(cluster, client)
-	state.RefreshClusterState()
-	state.RefreshContainerInstanceState()
-	state.RefreshTaskState()
-
-	clusterInstance := state.FindClusterByName(cluster)
+func StartTask(taskDefinitions []string) error {
+	// Cluster comes from global variable which is determined on the start of runtime of scheduler.
```
Comment (tnachen) [inline scheduler.go:64]: We need to refresh cluster state if it hasn't been updated for sometime.

Comment (swhsiang) [reply scheduler.go:64]: I would like to implement this function on next pull request by a job scheduler.
Perhaps add a parameter to decide the time interval of refreshing the cluster state. By default, the time interval is 5 mins.

File: api.go (hunk 3 above)
Comment (swhsiang) [reply api.go]: Here is my idea:

```
Register daemon ----POST----> daemon-ecs-scheduler ----> Start the daemon on each host immediately
                                          |
                                          |----> Every 5 minutes, the job scheduler checks whether each host is running the task (daemon) or not.
```

File: scheduler.go (hunk 4 above)
Comment (tnachen) [reply scheduler.go:64]: Hmm i think let's do it next. I think we should refresh more often than that, perhaps just 1 minute.

File: scheduler.go [hunk 5]
```diff
@@ -127,20 +129,27 @@ func main() {
...
 	}
 	app.Action = func(c *cli.Context) error {
-		if c.String("cluster") != "" {
-			taskDefinitions = c.StringSlice("task-definitions")
-			err := StartTask(cluster, taskDefinitions, awsRegion)
-			if err != nil {
-				return cli.NewExitError(err.Error(), 1)
-			}
+		if Cluster == "" {
+			return cli.NewExitError("cluster does exit", 1)
```
Comment (tnachen) [inline scheduler.go:141]: The error message doesn't make sense. I'll fix it for you, let's merge this.

File: scheduler.go [hunk 6]
```diff
@@ -127,20 +129,27 @@ func main() {
...
+		}
+		taskDefinitions = c.StringSlice("task-definitions")
+		if len(taskDefinitions) == 0 {
+			return cli.NewExitError("no tasks", 1)
```
Comment (tnachen) [inline scheduler.go:145]: We should always have a more descriptive error message.

**Pull request merged**